x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_list/credentials_list.test.tsx

```diff
@@ -0,0 +1,276 @@
+      expect(pagination).toEqual({
+        pageIndex: 5,
+        pageSize: 55,
+        totalItemCount: 1004,
+        hidePerPageOptions: true,
+      });
+    });
+
+    it('will default pagination values if `page` is not available', () => {
+      mockKea({
+        values: {
+          meta: {},
+        },
+      });
+
+      const wrapper = shallow(<CredentialsList />);
+      const { pagination } = wrapper.find(EuiBasicTable).props();
+      expect(pagination).toEqual({
+        pageIndex: 0,
+        pageSize: 0,
+        totalItemCount: 0,
+        hidePerPageOptions: true,
+      });
+    });
+  });
+
+  describe('columns', () => {
+    let columns: Array<{
+      render: (token: object) => any;
+      actions: any;
+    }>;
+    const showCredentialsForm = jest.fn();
+    const deleteApiKey = jest.fn();
+
+    beforeEach(() => {
+      mockKea({
+        actions: {
+          showCredentialsForm,
+          deleteApiKey,
+        },
+      });
+      const wrapper = shallow(<CredentialsList />);
+      columns = wrapper.find(EuiBasicTable).props().columns;
+    });
+
+    describe('column 1 (name)', () => {
+      const token = {
+        ...apiToken,
+        name: 'some-name',
+      };
+
+      it('renders correctly', () => {
+        const column = columns[0];
+        const wrapper = shallow(<div>{column.render(token)}</div>);
+        expect(wrapper.text()).toEqual('some-name');
+      });
+    });
+
+    describe('column 2 (type)', () => {
+      const token = {
+        ...apiToken,
+        type: ApiTokenTypes.Private,
+      };
+
+      it('renders correctly', () => {
+        const column = columns[1];
+        const wrapper = shallow(<div>{column.render(token)}</div>);
+        expect(wrapper.text()).toEqual('Private API Key');
+      });
+    });
+
+    describe('column 3 (key)', () => {
+      const testToken = {
+        ...apiToken,
+        key: 'abc-123',
+      };
+
+      const subject = (token: object): [any, any] => {
+        const copyMock = jest.fn();
+        const column = columns[2];
+        const wrapper = shallow(<div>{column.render(token)}</div>);
+        const children = wrapper.find(EuiCopy).props().children;
+        const copyEl = shallow(children(copyMock));
+        return [copyMock, copyEl];
+      };
+
+      it('renders the credential and a button to copy it', () => {
+        const [copyMock, copyEl] = subject(testToken);
+        expect(copyEl.find('EuiButtonIcon').props().onClick).toEqual(copyMock);
+        expect(copyEl.text().replace('<EuiButtonIcon />', '')).toBe('abc-123');
+      });
+
+      it('empty text if no key is present', () => {
+        const [copyMock, copyEl] = subject({
+          ...testToken,
+          key: undefined,
+        });
+        expect(copyEl.find('EuiButtonIcon').props().onClick).toEqual(copyMock);
+        expect(copyEl.text().replace('<EuiButtonIcon />', '')).toBe('');
+      });
+    });
```

Comment:
This test actually made me look at our original code a more closely. If we have no `apiToken.key`, should we even render the copy button at all? I know it should basically never happen, but it just seems odd to me to bother rendering the button if you're just copying an empty string. Should we consider going a little further and doing something like this?

```jsx

    {
      name: 'Key',
      width: '36%',
      render: (item: IApiToken) => {
        if (!item.key) return null;
        return (
          <EuiCopy
            textToCopy={item.key}
            afterMessage={i18n.translate('xpack.enterpriseSearch.appSearch.credentials.copied', {
              defaultMessage: 'Copied',
            })}
          >
            {(copy) => (
              <>
                <EuiButtonIcon
                  onClick={copy}
                  iconType="copyClipboard"
                  aria-label={i18n.translate(
                    'xpack.enterpriseSearch.appSearch.credentials.copyApiKey',
                    {
                      defaultMessage: 'Copy API Key to clipboard',
                    }
                  )}
                />
                {item.key}
              </>
            )}
          </EuiCopy>
        );
      },
    },
```

Reply:
Alternatively - if there is no realistic scenario where an API token does not have a `key` prop, I'd be +1 for just marking it as non-optional in Typescript and removing the `|| ''` branch

Reply:
I changed the branch to just do an early return: https://github.com/elastic/kibana/pull/79749/commits/a0ff6461eb37e6e3b2418785a5e73c1a8eefcc63?w=1

This all stems from the default API token set when the logic initializes. https://github.com/elastic/kibana/blob/master/x-pack/plugins/enterprise_search/public/applications/app_search/components/credentials/credentials_logic.ts#L18

I could change that key to initialize with an empty string. At this point tho I'm thinking I'd kind of like to just go forward with what we have here because we'd have to think through what other places in the UI that could affect.

Reply:
Ahh right right. Good call, early return works great for me as well 👍